### ISSUE TITLE
Query setting `max_memory_usage` doesn't work for `CREATE TABLE ... as SELECT ...`

### ISSUE THREAD

**Issue by alexsubota** [ClickHouse/ClickHouse#28810]
ClickHouse v.21.7.7.47

setting max_memory_usage doesnt work for 'create table .. as' queries

for example:
**this query works**
`create table test_table engine MergeTree order by a as
select a_table.a, b_table.b_arr
from (select arrayJoin(range(10000)) as a) a_table
         cross join (select range(10000) as b_arr) b_table
    settings max_memory_usage = 1;`

**this query doesnt work**
`insert into test_table
select a_table.a, b_table.b_arr
from (select arrayJoin(range(10000)) as a) a_table
         cross join (select range(10000) as b_arr) b_table
    settings max_memory_usage = 1;`
With error:
`DB::Exception: Memory limit (for query) exceeded: would use 4.09 MiB`

**Expected behavior**
**this query should NOT work, and setting max_memory_usage should work for all 'select'\'insert'\'create table' queries**
`create table test_table engine MergeTree order by a as
select a_table.a, b_table.b_arr
from (select arrayJoin(range(10000)) as a) a_table
         cross join (select range(10000) as b_arr) b_table
    settings max_memory_usage = 1;`

**Comment by alesapin** [comment]
Looks like it works for global settings:
```
click.sas.yp-c.yandex.net :) set max_memory_usage=1                                                                                                                                                                                                                              
                                                                                                                                                                                                                                                                                 
SET max_memory_usage = 1

Query id: f817742f-307b-4f55-b5e1-0c44c90c3ecb

Ok.

0 rows in set. Elapsed: 0.001 sec.
click.sas.yp-c.yandex.net :) create table test_table engine MergeTree order by a as select a_table.a, b_table.b_arr from (select arrayJoin(range(10000)) as a) a_table cross join (select range(10000) as b_arr) b_table settings max_memory_usage = 1;

CREATE TABLE test_table
ENGINE = MergeTree
ORDER BY a AS
SELECT
    a_table.a,
    b_table.b_arr
FROM
(
    SELECT arrayJoin(range(10000)) AS a
) AS a_table
CROSS JOIN
(
    SELECT range(10000) AS b_arr
) AS b_table

Query id: d1bdf4b8-f6c5-4df4-82e4-9bbb27b16427


0 rows in set. Elapsed: 0.038 sec.

Received exception from server (version 21.10.1):
Code: 241. DB::Exception: Received from localhost:9000. DB::Exception: Memory limit (for query) exceeded: would use 4.69 MiB (attempt to allocate chunk of 4915200 bytes), maximum: 1.00 B: While executing JoiningTransform. (MEMORY_LIMIT_EXCEEDED)

```

**Comment by alesapin** [comment]
But doesn't work for query settings:
```
click.sas.yp-c.yandex.net :) create table test_table engine MergeTree order by a as select a_table.a, b_table.b_arr from (select arrayJoin(range(10000)) as a) a_table cross join (select range(10000) as b_arr) b_table settings max_memory_usage = 1;

CREATE TABLE test_table
ENGINE = MergeTree
ORDER BY a AS
SELECT
    a_table.a,
    b_table.b_arr
FROM
(
    SELECT arrayJoin(range(10000)) AS a
) AS a_table
CROSS JOIN
(
    SELECT range(10000) AS b_arr
) AS b_table
SETTINGS max_memory_usage = 1

Query id: eb20341a-2daa-4c41-b36e-b6130ad205e9

Ok.

0 rows in set. Elapsed: 0.335 sec. 

```

**Comment by azat** [comment]
I've looked at this few days ago, and the problem is that those settings are not applied to the query context (since now query has only one context), #28962 should address this.

**Comment by oleg-savko** [comment]
Does it will be backported to previous version 21.3/21.8/21.7 ?

**Comment by azat** [comment]
>Does it will be backported to previous version 21.3/21.8/21.7 ?

Unlikely (AFAIU only major bug fixes are backported, @alexey-milovidov / @alesapin will correct me if I'm wrong)

**Comment by siradjev** [comment]
> Does it will be backported to previous version 21.3/21.8/21.7 ?
+1, it's causing unpredictable behavior for logic relying on settings, closer to a bug...

**Comment by oleg-savko** [comment]
@alexey-milovidov / @alesapin i think, at least it should be included in latest in 21.8-lts version)
